### PR TITLE
[SPARK-6079] Use index to speed up StatusTracker.getJobIdsForGroup()

### PR DESCRIPTION
**Pull request by JoshRosen**
`StatusTracker.getJobIdsForGroup()` is implemented via a linear scan over a HashMap rather than using an index, which might be an expensive operation if there are many (e.g. thousands) of retained jobs.

This patch adds a new map to `JobProgressListener` in order to speed up these lookups.

### CODE REVIEW

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
@pwendell @andrewor14, could one of you take a quick look at this patch?  Should be pretty straightforward.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala [hunk 1]
```diff
@@ -109,7 +111,8 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
       "failedJobs" -> failedJobs.size,
       "completedStages" -> completedStages.size,
       "skippedStages" -> skippedStages.size,
-      "failedStages" -> failedStages.size
+      "failedStages" -> failedStages.size,
+      "jobGroupToJobIds" -> jobGroupToJobIds.values.map(_.size).sum
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: IIUC, based on when elements are removed, the size of this should be the same as the size of `jobIdToData`.  Why not okace `jobGroupToJobIds` alongside it in `getSizesOfSoftSizeLimitedCollections`?

File: core/src/test/scala/org/apache/spark/ui/jobs/JobProgressListenerSuite.scala [hunk 2]
```diff
@@ -110,6 +120,23 @@ class JobProgressListenerSuite extends FunSuite with LocalSparkContext with Matc
...
+
+    // Run 50 jobs, each with one stage
+    for (jobId <- 0 to 50) {
+      listener.onJobStart(createJobStartEvent(jobId, Seq(0), jobGroup = Some(jobId.toString)))
+      listener.onStageSubmitted(createStageStartEvent(0))
+      listener.onStageCompleted(createStageEndEvent(0, failed = false))
+      listener.onJobEnd(createJobEndEvent(jobId, false))
+    }
+    assertActiveJobsStateIsEmpty(listener)
+    //
```
Comment (andrewor14) [inline core/src/test/scala/org/apache/spark/ui/jobs/JobProgressListenerSuite.scala]: ?

File: core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala [hunk 3]
```diff
@@ -140,7 +143,16 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
     if (jobs.size > retainedJobs) {
       val toRemove = math.max(retainedJobs / 10, 1)
       jobs.take(toRemove).foreach { job =>
-        jobIdToData.remove(job.jobId)
+        for (
+          removedJob <- jobIdToData.remove(job.jobId);
+          jobGroupId = removedJob.jobGroup.orNull;
+          jobsInGroup <- jobGroupToJobIds.get(jobGroupId)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: this mix and match of `<-` and `=` took me a while to figure out what's going on. Can we rewrite this:
```
jobs.take(toRemove).foreach { job =>
  jobIdToData.remove(job.jobId).foreach { removedJob =>
    val jobGroupId = removedJob.jobGroup.orNull
    val jobsInGroup = jobGroupToJobIds.get(jobGroupId)
    jobsInGroup.remove(job.jobId)
    ...
  }
}
```

File: core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala (hunk 1 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: Good point; I'll make this change.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala [hunk 4]
```diff
@@ -158,6 +170,7 @@ class JobProgressListener(conf: SparkConf) extends SparkListener with Logging {
         stageIds = jobStart.stageIds,
         jobGroup = jobGroup,
         status = JobExecutionStatus.RUNNING)
+    jobGroupToJobIds.getOrElseUpdate(jobGroup.orNull, new HashSet[JobId]).add(jobStart.jobId)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: what does it mean to have jobs in the `null` job group?

**Comment by andrewor14** [comment]
LGTM for the most part. I think this is mergeable as is but I would like to see the for loop written in a more readable way.

File: core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala (hunk 4 above)
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala]: The `null` job group contains jobs that do not have any other job group set (e.g. jobs without a job group).

**Comment by andrewor14** [comment]
Thanks for the new comments @JoshRosen. The latest version is really easy to follow. Merging into master.

**Comment by JoshRosen** [comment]
Also backported this to `branch-1.3` (1.3.1).

**Pull request closed without merging**